### PR TITLE
Adds fault tolerance design doc (do not merge until review is complete)

### PR DESCRIPTION
**Pull request by cnsnyder**
Please use this PR for review of this design.
For your convenience the latest version of this doc (with diagrams rendered etc) will be available [here](http://file.rdu.redhat.com/~csnyder/Candlepin%20-%20Fault_Tolerance_Design.html)

### CODE REVIEW

File: docs/candlepin/fault_tolerance_design.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
+---
+title: Fault_Tolerance_Design
+categories: Designs
+---
+{% include toc.md %}
+
+# PR for review
+Please use the PR located [here](https://github.com/candlepin/candlepinproject.org/pull/67)
+for any review and / or discussion of open questions or requirements
+
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: This section needs to be removed as we wouldn't want it in the final document.

File: docs/candlepin/fault_tolerance_design.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+## Problem
+
+Presently we have no way of tracking entitlements for each running instance of a virtual machine provided that each virtual machine has the same bios uuid. Tracking of entitlements is made even more difficult by technologies like vmware vLockstep etc that provide fault tolerant configurations. In these configurations there is a live backup vm that exactly duplicates the present state of the primary vm, meaning it cannot register to candlepin directly or have a different consumer_id. These fault tolerant configurations require the primary and backup vm to be on different hypervisors, meaning each vms compliance status could legitimately be quite different.
+
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Please wrap lines at 80 characters.  If you use vim, I can give you the settings to do this automatically.

File: docs/candlepin/fault_tolerance_design.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+### Requirements
+
+These are the requirements I have been designing around:
+
+- [Regarding two machines running in a fault tolerant configuration] From a business stand-point these are two different systems and each should be using a different subscription.
+- Reference [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1235727)
+- The compliance status of each instance should be independant of the status of other instances (one could be green in one place and red in another)
+- Each instance should be entitled based on it's hypervisor and consume the appropriate entitlements.
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Change `it's` to `its` as you require the possessive and not the contraction for `it is.`

File: docs/candlepin/fault_tolerance_design.md (hunk 2 above)
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: For formatting in vim, I have the following in my `~/.vimrc`

```
" Don't use Ex mode, use Q for formatting
map Q gq
" For all text files wrap at 80 characters and show right margin
autocmd FileType text,markdown setlocal textwidth=80 colorcolumn=80
" Opposite of 'J': split a line.  (By default 'K' runs man on the word under
" the cursor.)
nnoremap K i<CR><Esc>
```

You can quickly format a block of text with Q<motion>.  For example, Q} will wrap all the text at 80 characters down to the next blank line.  Or you can use K to insert a line break at the cursor position.

File: docs/candlepin/fault_tolerance_design.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+## Proposal
+
+Create a consumer to represent the live backup instance that, in essence, cannot directly register itself. We can detect the need for these consumers by tracking additional information provided by virt-who (described at length below). Failover can be detected by changes in this information. When this happens the primary consumer is labelled 'failedover'. When the fault tolerant guest machine next checks in or otherwise accesses candlepin claiming to be the failed over consumer candlepin will throw a new exception that will include the identity cert of the now current consumer. Python-rhsm will have to be modified to catch the exception and replace the present identity cert with the new one. After that point everything resumes normal operation.
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: * Capitalize Candlepin

File: docs/candlepin/fault_tolerance_design.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+## Necessary Changes
+
+In order to support Vmware Fault Tolerance, the following must be implemented:
+
+- virt-who should report the following s new attributes per guest:
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: `following s` should be `following`?

File: docs/candlepin/fault_tolerance_design.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    - vm.config.ftinfo.{role, instanceUUIDs}
+    - vm.config.instanceUUID
+
+- Candlepin should:
+    - Persist instanceUUIDs and associate them with consumers (stored as facts on the consumer that we only update server-side)
+    - Create a consumer per (guest_id, instanceUUID) on hypervisor checkin.
+        - Do not set the org for these consumers initial.
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: `initial` should be `initially`?

File: docs/candlepin/fault_tolerance_design.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        - The org for all of these consumers will be set based on what is submitted when the vm registers
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Capitalize VM

File: docs/candlepin/fault_tolerance_design.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+            - This implies that the set of all consumers that collectively represents this fault tolerant vm will all be part of the same org.
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: * I would rephrase this sentence as the antecedent of the pronoun `this` is unclear here.  Do you mean `this org setting implies...`?
* Capitalize VM
* Should `will all be` be `will be`?

File: docs/candlepin/fault_tolerance_design.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    - Detect (using the new report info) when a fail over has occured.
+        - Mark the original primary consumer (based on the ftinfo.role fact) as being a failed over state
+        - Provide the correct identity cert to use if a guest claims to be a failed over consumer in a new exception
+    - Check, on registration, if there is a consumer that exists already with the same virt.uuid as well as a ftinfo.role of 0, if so merge the info from registration into that consumer and return that one.
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: * Perhaps rephrase to `if there is an existing consumer with the same...`
* Either end the sentence at `0` or add a conjunction.  It's a run-on currently.
* Rephrase as `merge the info...into that consumer and return it`?

File: docs/candlepin/fault_tolerance_design.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    - Add the hypervisorId to the virt-only pools created
+        - This will simplify the logic of attaching to the pool provided for the consumer
+        - It also should allow sharing of such subs across orgs
+    - If there are hypervisors that provide subscriptions from different orgs (orgs other than the org to which all the consumers for this fault tolerant configuration are registered), then the subs they provide must be shared with the org the fault tolerant vm was registered to in order for them to be consumed.
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Capitalize VM here

File: docs/candlepin/fault_tolerance_design.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+- Subscription-manager / python-rhsm:
+    - Catch exceptions thrown by candlepin regarding failover detection.
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Capitalize Candlepin

File: docs/candlepin/fault_tolerance_design.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        - Entitlement certs should be refreshed after the new identity cert is in place.
+
+## How this affects Katello
+
+This section will include thoughts on how this will affect katello after this design is discussed with folks who work on Katello.
+
+## Open Questions
+
+- How do the appropriate subs get attached to the backup consumers? At this point we can detect and create consumers for the multiple instances of the same vm.
+- When we attach to a particular pool via the command line on the primary vm, what do we do for the back up vms? We cannot necessarily consume from the same pool as that contract might not have enough available to do so (but there might be a pool that provides similar products).
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Capitalize VM and change `vms` to `VMs`.

File: docs/candlepin/fault_tolerance_design.md [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+else (False)
+    :Proceed as normal;
+    stop
+endif
+
+{% endplantuml %}
+
+------------------------------------------------------------------------------------------------------
+
+
```
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Diagrams looks good!

File: docs/candlepin/fault_tolerance_design.md (hunk 4 above)
Comment (awood) [inline docs/candlepin/fault_tolerance_design.md]: Honestly, I'm not sure I'm that comfortable with Candlepin throwing an error/exception when everything is, in fact, okay.  What if we return a 205 Reset Content that python-rhsm/sub-man understand to mean, "I need to fetch a new identity certificate"?  Alternatively, some sort of 3xx code might be appropriate.  Using exceptions as control flow is kind of a code smell.

**Comment by awood** [comment]
What happens when the consumer that went kaput comes back online?

**Comment by cnsnyder** [comment]
@awood When an instance fails there is a new instance created with a different instanceUUID. Due to this behavior the original consumer (that went kaput) will never appear to come back online to us.

File: docs/candlepin/fault_tolerance_design.md (hunk 1 above)
Comment (cnsnyder) [inline docs/candlepin/fault_tolerance_design.md]: I'll make sure to remove this section once the design has been accepted.

**Comment by cnsnyder** [comment]
I think we have decided against implementation of this design. As such I'm closing this PR for now. If at some point later we decide to revisit, we can reopen this PR.
